File: 250926/그래프 탐색/graph-traversal.java
import java.util.*;
import java.io.*;

public class Main {
	
	static StringBuilder sb = new StringBuilder();
	static ArrayList<Integer>[] graph;
	static int n, m;
	static boolean[] visited;
	static int cnt;
	
	public static void main(String[] args) throws IOException {
		// System.setIn(new FileInputStream("src/coding_test/sample_input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		StringTokenizer st = new StringTokenizer(br.readLine().trim(), " ");
		n = Integer.parseInt(st.nextToken()); // 점 개수
		m = Integer.parseInt(st.nextToken()); // edge 개수
		
		graph = new ArrayList[n+1];
		for(int i =1; i < n+1; i ++) {
			graph[i] = new ArrayList<Integer>();
		}
		visited = new boolean[n+1]; // 기본 초기화 값 false
		
		// 그래프 세팅
		for(int i = 0; i < m; i ++) {
			st = new StringTokenizer(br.readLine().trim(), " ");
			int start = Integer.parseInt(st.nextToken());
			int to = Integer.parseInt(st.nextToken());
			
			graph[start].add(to);
			graph[to].add(start); // 방향 그래프가 아니라서 양쪽으로 더해줘야함
		}
		
		cnt = 0;
		visited[1] = true; // 출발점은 이미 방문했다고 표시해줘야함!
		DFS(1);
		
		bw.write(String.valueOf(cnt));
		bw.close();
	}
	
	public static void DFS(int vertex) {
		for(int i = 0; i < graph[vertex].size(); i++) {
			int currV = graph[vertex].get(i);
			// 연결된 점 아직 방문 안했으면
			if( !visited[currV] ) {
				// 조합, 순열처럼 다시 false로 표시할 필요x
				visited[currV] = true;
				cnt++;
				DFS(currV);
			}
		}
		
	}
}

File: 251016/그래프 탐색/graph-traversal.cpp
#include <iostream>
#include <vector>
using namespace std;

int n, m;
vector<int> visited;
vector<vector<int>> graph;

int dfs(int cur, int cnt){
    // cur 노드에 연결된 노드 리스트 순회
    for(int num : graph[cur]){
        if(visited[num] == 1)
            continue;
        visited[num] = 1; // 방문표시하기
        cnt++;
        cnt = dfs(num, cnt);
    }
    // visited 아닌 점을 모두 돌았으면 return
    return cnt;
}

int main() {

    // 1번 노드에서 시작해서 도달할 수 있는 서로다른 정점의 수
    cin >> n;
    cin >> m;

    visited.assign(n+1,0); // 점i를 방문했으면 visited[i]==1 

    //graph[i]는 점i에 연결된 점들 vector<int>
    graph.assign(n+1, vector<int>()); // 빈 벡터로 각 줄을 초기화 vector<int>()
    
    for(int i = 0; i < m; i++){
        int s, e;
        cin >> s;
        cin >> e;
        graph[s].push_back(e);
        graph[e].push_back(s);
    }

    visited[1] = 1;
    cout << dfs(1, 0);
    return 0;
}

